### PR TITLE
sstables: enable read ahead for the partition index reader

### PR DESCRIPTION
**Pull request by wmitros**
Currently, when advancing one of `index_reader`'s bounds, we're creating a new `index_consume_entry_context` with a new underlying file `input_stream` for each new page.

For either bound, the streams can be reused, because the indexes of pages that we are reading are never decreasing.

This patch adds a `index_consume_entry_context` to each of `index_reader`'s bounds, so that for each new page, the same file `input_stream` is used.
As a result, when reading consecutive pages, the reads that follow the first one can be satisfied by the `input_stream`'s read aheads, decreasing the number of blocking reads and increasing the throughput of the `index_reader`.

Additionally, we're reusing the `index_consumer` for all pages, calling `index_consumer::prepare` when we need to increase the size of  the `_entries` `chunked_managed_vector`.

A big difference can be seen when we're reading the entire table, frequently skipping a few rows; which we can test using perf_fast_forward:

Before:
```
running: small-partition-skips on dataset small-part
Testing scanning small partitions with skips.
Reads whole range interleaving reads with skips according to read-skip pattern:
   read    skip      time (s)   iterations     frags     frag/s    mad f/s    max f/s    min f/s    avg aio    aio      (KiB) blocked dropped  idx hit idx miss  idx blk    c hit   c miss    c blk    allocs   tasks insns/f    cpu 
-> 1       0         0.899447            4   1000000    1111794      12284    1113248    1096537      975.5    972     124356       1       0        0        0        0        0        0        0  12032202   29103    8967 100.0% 
-> 1       1         1.805811            4    500000     276884        907     278214     275977     3655.8   3654     135084    2688       0     3161     4548     5935        0        0        0   7225100  140466   27010  75.6% 
-> 1       8         0.927339            4    111112     119818        357     120465     119461     3654.0   3654     135084    2685       0     2133     4548     6963        0        0        0   1749663  107922   57502  50.2% 
-> 1       16        0.790630            4     58824      74401        782      74617      73497     3654.0   3654     135084    2695       0     1975     4548     7121        0        0        0   1019189  109349   90832  42.7% 
-> 1       32        0.717235            4     30304      42251        243      42266      41975     3654.0   3654     135084    2689       0     1871     4548     7225        0        0        0    619876  109199  156751  37.3% 
-> 1       64        0.681624            4     15385      22571        244      22815      22286     3654.0   3654     135084    2685       0     1870     4548     7226        0        0        0    407671  105798  285688  34.0% 
-> 1       256       0.630439            4      3892       6173         24       6214       6150     3549.0   3549     135116    2581       0     1313     3927     6505        0        0        0    232541  100803 1022454  29.1% 
-> 1       1024      0.313303            4       976       3115        219       3126       2766     1956.0   1956     130608     986       0        0      987     1962        0        0        0     81165   41385 1724979  29.1% 
-> 1       4096      0.083688            4       245       2928         85       3012       2134      738.8    737      17212     492     244        0      247      491        0        0        0     30500   19406 1999263  24.6% 
-> 64      1         1.509011            4    984616     652491       2746     660930     649745     3673.5   3654     135084    2687       0     4507     4548     4589        0        0        0  11075882  117074   13157  68.9% 
-> 64      8         1.424147            4    888896     624160       4446     625675     617713     3654.0   3654     135084    2691       0     4248     4548     4848        0        0        0  10019098  117383   13700  66.5% 
-> 64      16        1.343276            4    800000     595559       5834     605880     589725     3654.0   3654     135084    2698       0     3989     4548     5107        0        0        0   9043830  124022   14206  64.9% 
-> 64      32        1.249721            4    666688     533469       5056     536638     526212     3654.0   3654     135084    2688       0     3616     4548     5480        0        0        0   7570848  123043   15377  60.9% 
-> 64      64        1.154549            4    500032     433097      10215     443312     415001     3654.0   3654     135084    2703       0     3161     4548     5935        0        0        0   5718758  110657   17787  53.2% 
-> 64      256       1.005309            4    200000     198944       1179     199338     196989     3935.0   3935     137216    2966       0      690     4048     5592        0        0        0   2398359  110510   27855  51.3% 
-> 64      1024      0.441913            4     58880     133239       8094     135471     120467     2161.0   2161     131820    1190       0        0     1192     1848        0        0        0    725092   45449   33740  59.7% 
-> 64      4096      0.124826            4     15424     123564       5958     126814      95101      795.5    794      17400     553     240        0      312      482        0        0        0    199943   20869   46621  41.9% 
```
After:
```
running: small-partition-skips on dataset small-part
Testing scanning small partitions with skips.
Reads whole range interleaving reads with skips according to read-skip pattern:
   read    skip      time (s)   iterations     frags     frag/s    mad f/s    max f/s    min f/s    avg aio    aio      (KiB) blocked dropped  idx hit idx miss  idx blk    c hit   c miss    c blk    allocs   tasks insns/f    cpu 
-> 1       0         0.917468            4   1000000    1089956       1422    1091378    1073112      975.5    972     124356       1       0        0        0        0        0        0        0  12032761   29721    8972 100.0% 
-> 1       1         1.311446            4    500000     381259       3212     384470     377238     1087.0   1083     138420       2       0     4445     4548     4651        0        0        0   7096216   55681   20869 100.0% 
-> 1       8         0.467975            4    111112     237432       1446     239372     235985     1121.2   1119     143124       9       0     4344     4548     4752        0        0        0   1619944   23502   28844  98.7% 
-> 1       16        0.337085            4     58824     174508       3410     178451     171099     1117.5   1120     143276      11       0     4319     4548     4777        0        0        0    883692   19152   37460  96.8% 
-> 1       32        0.262798            4     30304     115313       1222     116535     112400     1070.2   1066     135620     166      26     4354     4548     4742        0        0        0    483185   18856   54275  94.9% 
-> 1       64        0.283954            4     15385      54181        531      56177      53650     2022.5   2040     137036     319      19     4351     4548     4745        0        0        0    292766   32998  102276  84.9% 
-> 1       256       0.207020            4      3892      18800        575      19105      17520     1315.5   1334     136072     418      24     3703     3927     4115        0        0        0    118400   27427  292146  82.1% 
-> 1       1024      0.164396            4       976       5937         57       5993       5842     1208.2   1195     135384     568      14      932      987     1030        0        0        0     62999   27554  503559  70.0% 
-> 1       4096      0.085079            4       245       2880        108       2987       2714      635.8    634      26468     248     246      233      247      258        0        0        0     31264   12872 1546404  37.4% 
-> 64      1         1.073331            4    984616     917346       7614     923983     909314     1812.2   1824     136792      11      20     4544     4548     4552        0        0        0  10971661   54538    9919  99.6% 
-> 64      8         1.024389            4    888896     867733       6327     870429     845215     3027.2   3072     138212      31       0     4523     4548     4573        0        0        0   9933078   68059   10050  99.5% 
-> 64      16        0.978754            4    800000     817366       7802     827665     809564     3012.2   3008     139884      39       0     4486     4548     4610        0        0        0   8947041   64050   10302  98.1% 
-> 64      32        0.837266            4    666688     796267      10312     806579     785370     2275.8   2266     139672      29       0     4465     4548     4631        0        0        0   7458644   50754   10564  97.8% 
-> 64      64        0.645627            4    500032     774490       4713     779203     768432     1136.8   1137     145428       8       0     4438     4548     4658        0        0        0   5593168   29982   10938  98.4% 
-> 64      256       0.386192            4    200000     517877      22509     544067     495368     1134.8   1136     145300     109       0     2135     4048     4147        0        0        0   2270291   22840   13682  94.5% 
-> 64      1024      0.238617            4     58880     246755      55856     305110     190899     1176.0   1118     135324     451      13      625     1192     1223        0        0        0    701262   24418   17323  71.1% 
-> 64      4096      0.133340            4     15424     115674      14837     117978      99072      974.0    961      27132     366     347       99      312      383        0        0        0    209595   20657   43096  50.4% 
```
For single partition reads, the index_reader is modified to behave in practically the same way, as before the change (not reading ahead past the page with the partition).
For example, a single partition read from a table with 10 rows per partition performs a single 6KB read from the index file, and the same read is performed before the change (as can be seen in traces below). If we enabled read aheads in that case, we would perform 2 16KB reads.
Relevant traces:
Before:
```
./tmp/data/ks/t2-75ebed30eb0211eb837a8f4cd3d1cf62/md-1-big-Index.db: scheduling bulk DMA read of size 6478 at offset 0 [shard 0] | 2021-07-23 15:22:25.847362 | 127.0.0.1 |            148 | 127.0.0.1
./tmp/data/ks/t2-75ebed30eb0211eb837a8f4cd3d1cf62/md-1-big-Index.db: finished bulk DMA read of size 6478 at offset 0, successfully read 6478 bytes [shard 0] | 2021-07-23 15:22:25.900996 | 127.0.0.1 |          53782 | 127.0.0.1
```
After:
```
./tmp/data/ks/t2-75ebed30eb0211eb837a8f4cd3d1cf62/md-1-big-Index.db: scheduling bulk DMA read of size 6478 at offset 0 [shard 0] | 2021-07-23 15:19:37.380033 | 127.0.0.1 |            149 | 127.0.0.1
./tmp/data/ks/t2-75ebed30eb0211eb837a8f4cd3d1cf62/md-1-big-Index.db: finished bulk DMA read of size 6478 at offset 0, successfully read 6478 bytes [shard 0] | 2021-07-23 15:19:37.433662 | 127.0.0.1 |          53777 | 127.0.0.1
```
Tests: unit(dev)

### CODE REVIEW

**Comment by denesb** [comment]
> Currently, when advancing one of `index_reader`'s bounds, we're creating a new `index_consume_entry_context` with a new underlying file `input_stream` for each new page.
> 
> For either bound, the streams can be reused, because the indexes of pages that we are reading are never decreasing.

This might be the case today but we have plans to change this to support [reverse reads](https://github.com/scylladb/scylla/issues/1413). What I want to say is that although it is fine to build on this assumption to speed up the common fast path of forward reads, we should not make reading in a zig-zag pattern impossible. I am saying this without having reviewed the code, so this might be a moot point, but wanted to state it nevertheless.

**Comment by avikivity** [comment]
I can't say I can follow the intricacies of the index reader, but overall looks good, especially the results.

Q: how does this affect single small partition read? Will it cause larger I/Os on the index?

I think you can obtain an answer with tracing, it will show file offsets and sizes on which we initate I/O.

**Comment by avikivity** [comment]
@tgrabiec please review

**Comment by wmitros** [comment]
> > Currently, when advancing one of `index_reader`'s bounds, we're creating a new `index_consume_entry_context` with a new underlying file `input_stream` for each new page.
> > For either bound, the streams can be reused, because the indexes of pages that we are reading are never decreasing.
> 
> This might be the case today but we have plans to change this to support [reverse reads](https://github.com/scylladb/scylla/issues/1413). What I want to say is that although it is fine to build on this assumption to speed up the common fast path of forward reads, we should not make reading in a zig-zag pattern impossible. I am saying this without having reviewed the code, so this might be a moot point, but wanted to state it nevertheless.

You're right that this patch is mostly about speeding up the common fast path, but I'm also going to work on the reverse reads, so I'll add my 2 cents about the interaction:
As far as I know, we are going to focus on optimizing the reversed reads from a single partition, and the streams that this patch combines are used for reading the partition data, not the rows in a single partition (for that we have promoted index, which has a separate stream).
Besides, I haven't read much into the reasons why, but the index reader already assumes that we're reading pages in a non decreasing order as can be seen here: https://github.com/scylladb/scylla/blob/7261c2c93e633a5df2772cc1313da59712ba74d4/sstables/index_reader.hh#L485-L488 so it is not a new assumption in this patch

**Comment by avikivity** [comment]
> As far as I know, we are going to focus on optimizing the reversed reads from a single partition, and the streams that this patch combines are used for reading the partition data, not the rows in a single partition (for that we have promoted index, which has a 

Reversed reads are only defined within a single partition. This is because multi-partition reads return partitions in token order, which is arbitrary as far as the application is concerned, whereas clustering key order is meaningful.

**Comment by wmitros** [comment]
> > As far as I know, we are going to focus on optimizing the reversed reads from a single partition, and the streams that this patch combines are used for reading the partition data, not the rows in a single partition (for that we have promoted index, which has a
> 
> Reversed reads are only defined within a single partition. This is because multi-partition reads return partitions in token order, which is arbitrary as far as the application is concerned, whereas clustering key order is meaningful.

Right, though we do allow reads with reversed order on clustering key for multiple partitions (WHERE pk IN ...), with paging disabled, but I assume after we implement reversed reads for single partition, we can perform the same process for multiple partitions

**Comment by wmitros** [comment]
> I can't say I can follow the intricacies of the index reader, but overall looks good, especially the results.
> 
> Q: how does this affect single small partition read? Will it cause larger I/Os on the index?
> 
> I think you can obtain an answer with tracing, it will show file offsets and sizes on which we initate I/O.

I'll try to test it soon, but before that I think I should add that there is one improvement to this patch that I've attempted but didn't seem worth the extra code churn, which would help with decreasing the amount of I/Os on the index:
In this patch, in order to enable advancing the index read context, the underlying file input stream is set to cover the entire index file. The improvement is that we could limit it when we know that the associated sstable reader is not forwarding - in that case we would only need to set the limits of the input stream to the partition range of the sstable reader

**Comment by avikivity** [comment]
> > > As far as I know, we are going to focus on optimizing the reversed reads from a single partition, and the streams that this patch combines are used for reading the partition data, not the rows in a single partition (for that we have promoted index, which has a
> > 
> > 
> > Reversed reads are only defined within a single partition. This is because multi-partition reads return partitions in token order, which is arbitrary as far as the application is concerned, whereas clustering key order is meaningful.
> 
> Right, though we do allow reads with reversed order on clustering key for multiple partitions (WHERE pk IN ...), with paging disabled, but I assume after we implement reversed reads for single partition, we can perform the same process for multiple partitions

I imagine it returns the partitions in the same order as unreversed reads. So even if we implement it for multiple partitions (for which there is no reason to, at least yet), we'll probably do the same.

I can see some use for 

```sql
SELECT *
FROM tab
ORDER BY score DESC
LIMIT 10
```

to give the best 10 scores per partition. Nobody cares about the partition order, but the clustering key order (assuming score is a clustering key here, which is dubious) is important.

**Comment by avikivity** [comment]
> > I can't say I can follow the intricacies of the index reader, but overall looks good, especially the results.
> > Q: how does this affect single small partition read? Will it cause larger I/Os on the index?
> > I think you can obtain an answer with tracing, it will show file offsets and sizes on which we initate I/O.
> 
> I'll try to test it soon, but before that I think I should add that there is one improvement to this patch that I've attempted but didn't seem worth the extra code churn, which would help with decreasing the amount of I/Os on the index:
> In this patch, in order to enable advancing the index read context, the underlying file input stream is set to cover the entire index file. The improvement is that we could limit it when we know that the associated sstable reader is not forwarding - in that case we would only need to set the limits of the input stream to the partition range of the sstable reader

Hmm. For multi-partition scans, it is not hugely important. We can expect that the ranges are usually large.

**Comment by wmitros** [comment]
> I can't say I can follow the intricacies of the index reader, but overall looks good, especially the results.
> 
> Q: how does this affect single small partition read? Will it cause larger I/Os on the index?
> 
> I think you can obtain an answer with tracing, it will show file offsets and sizes on which we initate I/O.

Ok, so after a little testing I can share my findings. First of all the reader always starts with 2 16KB reads, because it has `read_ahead` and `dynamic_adjustments` in its `file_input_stream_options`, and the reader (as in the previous version) reads an entire page (from summary) - if the page is bigger than 16KB we have 2 reads for the entire page, and 1 read-ahead.
The page size is proportional to the partition key size, and inversely proportional to the average number of rows in a partition.

Assuming we aren't using huge partition keys we are unlikely to read the page with more than one read (I only got such big pages with ~1 row per partition), so it should be safe to assume we have 2 16KB reads for single small partition reads.

Before the changes, we were also reading one partition, and despite setting `read_ahead` in `file_input_stream_options`, we weren't really making read-aheads, because the file stream was limited to the current page. With one row per partition, we got one 16KB read, and one 6-10KB reads, with 10 rows per partition we got one 4-6KB read, for 100 rows we still had more single reads of more than 1KB.

If the single small partition reads are common, we can fix them with some more ifs and with passing an additional bool to the index read context code, though I can't tell if that's worth it.

**Comment by avikivity** [comment]
Single small partition reads are extremely common. More common than anything else.

But, if this series does not regress single small partition reads, we can fix it in a separate patch. Not only to reduce read-ahead, but also to clamp the read size to what the summary entries tell us (could mean we'd want to read more than 16k).

btw, there's room for improvement in how summary selectivity works. With low disk:ram ratios, we could make it more selective (and so have to read less from index) at the cost of more memory footprint. With high disk:ram ratios, the inverse.

```c++
void maybe_add_summary_entry(summary& s, const dht::token& token, bytes_view key, uint64_t data_offset,
        uint64_t index_offset, index_sampling_state& state) {
    state.partition_count++;
    // generates a summary entry when possible (= keep summary / data size ratio within reasonable limits)
    if (data_offset >= state.next_data_offset_to_write_summary) {
        auto entry_size = 8 + 2 + key.size();  // offset + key_size.size + key.size
        state.next_data_offset_to_write_summary += state.summary_byte_cost * entry_size;
        s.add_summary_data(token.data());
        auto key_data = s.add_summary_data(key);
        s.entries.push_back({ token, key_data, index_offset });
    }
}

// Returns the cost for writing a byte to summary such that the ratio of summary
// to data will be 1 to cost by the time sstable is sealed.
size_t summary_byte_cost(double summary_ratio) {
    return summary_ratio ? (1 / summary_ratio) : index_sampling_state::default_summary_byte_cost;
}
```

**Comment by avikivity** [comment]
(and we can stop storing partition keys in the in-memory summary, just on disk, and so make the summary even more selective)

**Comment by wmitros** [comment]
The change needed for making single partition reads as small as before the change turned out to be pretty small, so I've added a commit with it

**Comment by avikivity** [comment]
Please add some test results to show the single-partition case to the cover letter (can be output of tracing).

I still need @tgrabiec 's review.

**Comment by wmitros** [comment]
> Please add some test results to show the single-partition case to the cover letter (can be output of tracing).
> 
> I still need @tgrabiec 's review.

Ok, I've added traces for an example single partition read from table with 10 rows per partition

**Comment by haaawk** [comment]
@tgrabiec could you have a look please?

**Comment by haaawk** [comment]
@tgrabiec @avikivity ping

**Comment by tgrabiec** [comment]
@haaawk It's in my queue.

File: sstables/index_reader.hh [hunk 1]
```diff
@@ -526,6 +527,14 @@ private:
                     if (ex) {
                         return make_exception_future<index_list>(std::move(ex));
                     }
+                    if (_single_page_read) {
+                        // if the associated reader is forwarding despite having singular range, we prepare for that
+                        _single_page_read = false;
+                        auto moved_context = std::move(bound.context);
+                        return moved_context->close().then([this] {
```
Comment (tgrabiec) [inline sstables/index_reader.hh:534]: moved_context can go out of scope before close() completes, potential use after free

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: sstables/index_reader.hh [hunk 2]
```diff
@@ -301,9 +301,9 @@ public:
     }
 
     index_consume_entry_context(reader_permit permit, IndexConsumer& consumer, trust_promoted_index trust_pi, const schema& s,
-            file index_file, file_input_stream_options options, uint64_t start,
-            uint64_t maxlen, std::optional<column_values_fixed_lengths> ck_values_fixed_lengths, tracing::trace_state_ptr trace_state = {})
-        : continuous_data_consumer(std::move(permit), make_file_input_stream(index_file, start, maxlen, options), start, maxlen)
+            input_stream<char> && input, uint64_t start, uint64_t maxlen,
```
[the remark this answers is not in the document]
Comment (wmitros) [reply sstables/index_reader.hh]: Ok

File: sstables/index_reader.hh [hunk 3]
```diff
@@ -442,32 +443,26 @@ class index_reader {
...
+        auto index_file = make_tracked_index_file(*_sstable, _permit, _trace_state, _use_caching);
+        auto input = make_file_input_stream(index_file, begin, _sstable->index_size() - begin,
+                        get_file_input_stream_options(_pc));
+        auto trust_pi = trust_promoted_index(_sstable->has_correct_promoted_index_entries());
+        auto ck_values_fixed_lengths = _sstable->get_version() >= sstable_version_types::mc
+                            ? std::make_optional(get_clustering_values_fixed_lengths(_sstable->get_serialization_header()))
+                            : std::optional<column_values_fixed_lengths>{};
+        return std::make_unique<index_consume_entry_context<index_consumer>>(_permit, _consumer, trust_pi,*_sstable->_schema, std::move(input),
+                            begin, end - begin, ck_values_fixed_lengths, _trace_state);
+    }
```
[the remark this answers is not in the document]
Comment (wmitros) [reply sstables/index_reader.hh:459]: Ok

File: sstables/index_reader.hh (hunk 1 above)
Comment (wmitros) [reply sstables/index_reader.hh]: I've added a `do_with` to prevent it in the new version

**Comment by wmitros** [comment]
I've addressed the review comments in the last rebase. Hopefully it's ready for merge now.
Sorry for fixing them so late, I didn't see the review initially and later forgot about it

**Comment by haaawk** [comment]
@avikivity @tgrabiec please have a look

File: sstables/index_reader.hh [hunk 4]
```diff
@@ -527,6 +528,15 @@ private:
                     if (ex) {
                         return make_exception_future<index_list>(std::move(ex));
                     }
+                    if (_single_page_read) {
+                        // if the associated reader is forwarding despite having singular range, we prepare for that
+                        _single_page_read = false;
+                        return do_with(std::move(bound.context), [this] (std::unique_ptr<index_consume_entry_context<index_consumer>>& moved_context) {
```
Comment (tgrabiec) [inline sstables/index_reader.hh:534]: We could avoid the allocation incurred by do_with():

```
auto& ctx = *bound.context;
return ctx.close().then([bc = std::move(bound.context)] {});
```

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: sstables/index_reader.hh (hunk 4 above)
Comment (wmitros) [reply sstables/index_reader.hh]: Ok

**Comment by wmitros** [comment]
The last rebase contained only the fix for the last review comment

**Comment by wmitros** [comment]
@tgrabiec I've applied your suggestion, please check if everything is fine now

**Comment by avikivity** [comment]
Dequeued, tests fail in release mode:

```console
$ ./test.py --mode release sstable_3_x_test --repeat 100
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[59/9000]   boost   release [ PASS ] sstable_3_x_test.test_write_overlapped_start_range_tombstones.69                                    
[60/9000]   boost   release [ FAIL ] sstable_3_x_test.test_read_rows_only_index.73
[5639/9000]  boost   release [ PASS ] sstable_3_x_test.test_write_adjacent_range_tombstones_with_rows.5646                                  
[5640/9000]  boost   release [ FAIL ] sstable_3_x_test.test_read_rows_only_index.5653
[5820/9000]  boost   release [ PASS ] sstable_3_x_test.test_write_range_tombstone_same_start_with_row.5827                                  
[5821/9000]  boost   release [ FAIL ] sstable_3_x_test.test_read_rows_only_index.5833
[5905/9000]  boost   release [ PASS ] sstable_3_x_test.test_write_compact_table.5909                                                        
[5906/9000]  boost   release [ FAIL ] sstable_3_x_test.test_read_rows_only_index.5923
[6195/9000]  boost   release [ PASS ] sstable_3_x_test.test_write_many_partitions_lz4.6168^C                                                
Shutdown requested... Aborting tests:
```

**Comment by haaawk** [comment]
@wmitros please have a look.

**Comment by haaawk** [comment]
Apparently it's an issue with the test and @bhalevy has already fixed it.

**Comment by avikivity** [comment]
@haaawk reference?

@wmitros once the test fix promotes, please rebase, so we keep the tree bisectable for tests too

**Comment by avikivity** [comment]
ah, probably 2f2e3b2e84eef3410b5834c4a394fc60d50521ac

**Comment by wmitros** [comment]
I couldn't replicate fails in the tests, but the alleged fix has been promoted, so I've uploaded the rebased version

**Comment by bhalevy** [comment]
> I couldn't replicate fails in the tests, but the alleged fix has been promoted, so I've uploaded the rebased version

yeah, it reproduced about 1/1000 runs for me

**Comment by avikivity** [comment]
Dequeued, fails cdc_test.py::TestCdc::test_cluster_expansion_with_cdc. @bhalevy may have more details.

**Comment by bhalevy** [comment]
With this change, `cdc_test.py::TestCdc::test_cluster_expansion_with_cdc` trips https://github.com/scylladb/scylla/blob/f7b8b809d063807096a11fecb797c334db266306/sstables/index_reader.hh#L548-L550
in high probability.  I didn't figure out why exactly.  Hopefully, @wmitros, if you can please reproduce this locally and investigate, this might also help us figure out a possibly related bug that @denesb is working on: #9446

Cc @tgrabiec

**Comment by tgrabiec** [comment]
Maybe `_single_page_read=true` index reader is advanced to the next partition.

**Comment by wmitros** [comment]
Looks like the last version doesn't fail on `cdc_test.py::TestCdc::test_cluster_expansion_with_cdc` any more (passed all tries when running for 12h at https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1378/), it also passes gating so we can try merging it again (https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1380/).

The issue was that previously I was using a single `index_consumer` for every `index_reader`, so it was common for both `_lower_bound` and `_upper_bound`. It wouldn't be a problem if used synchronously, but we have https://github.com/scylladb/scylla/blob/680195564de7a54f5a6b2317a648a8e5671e1afe/sstables/index_reader.hh#L769-L773 where consumer would consume from both bounds at the same time. This was fixed by moving the consumer to the bounds, so that each bound has its own consumer

@avikivity @tgrabiec please take a look

**Comment by bhalevy** [comment]
@tgrabiec can you please re-review? See https://github.com/scylladb/scylla/pull/9063#issuecomment-1049701354

**Review by tgrabiec — APPROVED** [review]

**Pull request closed without merging**